After some time plugin backend calls never resolve and decky-loader goes into a weird state

Hi,

while writing a new plugin for taking notes : https://github.com/popsUlfr/SDH-Notebook I've run into the issue where calls to the python backend never return after some time. I'm not sure if I did something very wrong in my code but once it happens the other plugins fail to function properly. decky-loader itself for instance is unable to check for updates and stays in a perpetual `Updates` spinner state :

![image](https://user-images.githubusercontent.com/1030423/187076184-dc85bb4b-6235-4f2f-b270-de97aa05ba96.png)

* decky-loader version: `v2.0.5-pre18`

Restarting `plugin_loader.service` seems to fix it until it happens again.

I added timeout checks to my backend calls and it seems to be stuck at the `importReactPlugin()` step:

```js
async importPlugin(name, version) {
    if (this.reloadLock) {
        this.log('Reload currently in progress, adding to queue', name);
        this.pluginReloadQueue.push({ name, version: version });
        return;
    }
    try {
        this.reloadLock = true;
        this.log(`Trying to load ${name}`);
        this.unloadPlugin(name);
        if (name.startsWith('$LEGACY_')) {
            await this.importLegacyPlugin(name.replace('$LEGACY_', ''));
        }
        else {
            await this.importReactPlugin(name, version); // <= this call
        }
        this.deckyState.setPlugins(this.plugins);
        this.log(`Loaded ${name}`);
    }
    catch (e) {
        throw e;
    }
    finally {
        this.reloadLock = false;
        const nextPlugin = this.pluginReloadQueue.shift();
        if (nextPlugin) {
            this.importPlugin(nextPlugin.name, nextPlugin.version);
        }
    }
}
async importReactPlugin(name, version) {
    let res = await fetch(`http://127.0.0.1:1337/plugins/${name}/frontend_bundle`, {
        credentials: 'include',
        headers: {
            Authentication: window.deckyAuthToken,
        },
    });
    if (res.ok) {
        let plugin = await eval(await res.text())(this.createPluginAPI(name)); // <= blocks here
        this.plugins.push({
            ...plugin,
            name: name,
            version: version,
        });
    }
    else
        throw new Error(`${name} frontend_bundle not OK`);
}
```

```
/8501.js:2 Error: Backend call timed out
    at eval (eval at importReactPlugin (localhost:1337/frontend/index.js:1125), <anonymous>:19187:33)
    at eval (eval at importReactPlugin (localhost:1337/frontend/index.js:1125), <anonymous>:19544:3)
    at PluginLoader.importReactPlugin (localhost:1337/frontend/index.js:1125)
    at async PluginLoader.importPlugin (localhost:1337/frontend/index.js:1101)
console.error @ /8501.js:2
```

EDIT1:
I just noticed this in the plugin_loader logs:

```
Aug 28 16:24:14 steamdeck PluginLoader[5871]: [base_events][ERROR]: Task exception was never retrieved
Aug 28 16:24:14 steamdeck PluginLoader[5871]: future: <Task finished name='Task-64' coro=<PluginWrapper._listen_for_method_call() done, defined at plugin.py:67> exception=ValueError('Separator is found, but chunk is longer than limit')>
Aug 28 16:24:14 steamdeck PluginLoader[5871]: Traceback (most recent call last):
Aug 28 16:24:14 steamdeck PluginLoader[5871]:   File "asyncio/streams.py", line 525, in readline
Aug 28 16:24:14 steamdeck PluginLoader[5871]:   File "asyncio/streams.py", line 620, in readuntil
Aug 28 16:24:14 steamdeck PluginLoader[5871]: asyncio.exceptions.LimitOverrunError: Separator is found, but chunk is longer than limit
Aug 28 16:24:14 steamdeck PluginLoader[5871]: During handling of the above exception, another exception occurred:
Aug 28 16:24:14 steamdeck PluginLoader[5871]: Traceback (most recent call last):
Aug 28 16:24:14 steamdeck PluginLoader[5871]:   File "plugin.py", line 69, in _listen_for_method_call
Aug 28 16:24:14 steamdeck PluginLoader[5871]:   File "asyncio/streams.py", line 534, in readline
Aug 28 16:24:14 steamdeck PluginLoader[5871]: ValueError: Separator is found, but chunk is longer than limit
```

So it would be an issue with the string length in this particular use case ?

--- a/backend/plugin.py
+++ b/backend/plugin.py
@@ -1,7 +1,7 @@
 import multiprocessing
 from asyncio import (Lock, get_event_loop, new_event_loop,
                      open_unix_connection, set_event_loop, sleep,
-                     start_unix_server)
+                     start_unix_server, IncompleteReadError, LimitOverrunError)
 from concurrent.futures import ProcessPoolExecutor
 from importlib.util import module_from_spec, spec_from_file_location
 from json import dumps, load, loads
@@ -66,7 +66,19 @@ class PluginWrapper:
 
     async def _listen_for_method_call(self, reader, writer):
         while True:
-            data = loads((await reader.readline()).decode("utf-8"))
+            line = bytearray()
+            while True:
+                try:
+                    line.extend(await reader.readuntil())
+                except LimitOverrunError:
+                    line.extend(await reader.read(reader._limit))
+                    continue
+                except IncompleteReadError as err:
+                    line.extend(err.partial)
+                    break
+                else:
+                    break
+            data = loads(line.decode("utf-8"))
             if "stop" in data:
                 get_event_loop().stop()
                 while get_event_loop().is_running():
@@ -121,7 +133,19 @@ class PluginWrapper:
                 self.writer.write(
                     (dumps({"method": method_name, "args": kwargs})+"\n").encode("utf-8"))
                 await self.writer.drain()
-                res = loads((await self.reader.readline()).decode("utf-8"))
+                line = bytearray()
+                while True:
+                    try:
+                        line.extend(await self.reader.readuntil())
+                    except LimitOverrunError:
+                        line.extend(await self.reader.read(self.reader._limit))
+                        continue
+                    except IncompleteReadError as err:
+                        line.extend(err.partial)
+                        break
+                    else:
+                        break
+                res = loads(line.decode("utf-8"))
                 if not res["success"]:
                     raise Exception(res["res"])
                 return res["res"]

--- a/backend/plugin.py
+++ b/backend/plugin.py
@@ -12,6 +12,8 @@ from time import time
 
 multiprocessing.set_start_method("fork")
 
+BUFFER_LIMIT = 2 ** 20  # 1 MiB
+
 class PluginWrapper:
     def __init__(self, file, plugin_directory, plugin_path) -> None:
         self.file = file
@@ -62,7 +64,7 @@ class PluginWrapper:
         get_event_loop().run_forever()
 
     async def _setup_socket(self):
-        self.socket = await start_unix_server(self._listen_for_method_call, path=self.socket_addr)
+        self.socket = await start_unix_server(self._listen_for_method_call, path=self.socket_addr, limit=BUFFER_LIMIT)
 
     async def _listen_for_method_call(self, reader, writer):
         while True:
@@ -100,7 +102,7 @@ class PluginWrapper:
             retries = 0
             while retries < 10:
                 try:
-                    self.reader, self.writer = await open_unix_connection(self.socket_addr)
+                    self.reader, self.writer = await open_unix_connection(self.socket_addr, limit=BUFFER_LIMIT)
                     return True
                 except:
                     await sleep(2)

